gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcCallContext.java

```diff
@@ -206,7 +206,10 @@ public ApiCallContext merge(ApiCallContext inputCallContext) {
     }
 
     CallOptions newCallOptions =
-        this.callOptions.withCallCredentials(newCallCredentials).withDeadline(newDeadline);
+        grpcCallContext
```

Comment:
I wanted to flag this change specifically. Is this the right thing to do? The result is that newCallOptions will now have the custom options in the inputCallContext.callOptions instead of the ones from this.callOptions.

Reply:
Hm... A good question. When I try to trace usages of `this.callOptions`, it seems it is used (most heavily in tests), but everytime it is initialized with DEFAULT options (which is empty). So, it looks like current stuff will still work if you leave things as is (since merging empty options somewhere is equal to not merging anything). 

This method (where the change is) is called "merge", so probably correct way of doing it is to merge `grpcCallContext.callOptions` and `this.callOptions` into a new "big" options object. It is just my speculations, but it seems like this method had a bug on merge: it was ignoring `grpcCallContext.callOptions` when merging (using only `this.callOptions`). Your change "flipped" the bug: now it still ignores one of 2 options, but this time it is `this.callOptions` which are ignored. 

We also probably need a test for this behavior.

Reply:
Unfortunately leaving things as they are does not work, because of this call: https://github.com/googleapis/gax-java/blob/master/gax/src/main/java/com/google/api/gax/rpc/UnaryCallable.java#L126 With the current beahviour, the custom call options are lost.

You are right that this change just "flips" the bug. Unfortunately there is a problem - the CallOptions class does not provide a way to extract all keys. The only available methods are `withOption` and `getOptions`. So I don't know if there is any way for us to implement merging of `CallOption` objects in a way that preserves the custom options in both objects.

Given that - what is the correct direction for preservation? Or, do we need to discard the merging approach entirely?

Reply:
I think we need to preserve the merge behavior; it's an important part of the semantics of our library, because it allows us to default everything and allow users to override only what they need to.

I think as for relevancy, our library won't initialize anything in `CallOptions`; I think the only way this case is relevant is when the user independently constructs two instances of `GrpcCallContext`, sets the `CallOptions` on each, and then tries to merge one into the other. 

Could you file an issue against grpc-java asking how to do such a merge, and if it's not possible, ask for the functionality?

Reply:
Filed https://github.com/grpc/grpc-java/issues/4251